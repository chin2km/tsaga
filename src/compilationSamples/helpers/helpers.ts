import * as path from 'path';
// tslint:disable-next-line:no-implicit-dependencies (Just used in tests)
import * as ts from 'typescript';

interface ReducedDiagnostic {
  message: string;
  fileName?: string;
  line?: number;
  character?: number;
}

export const defaultCompilerOptions: ts.CompilerOptions = {
  noEmit: true,
  strict: true,
  target: ts.ScriptTarget.ES2016,
  module: ts.ModuleKind.CommonJS,
  jsxFactory: 'react',
};

export function reducedDiagnostic(
  diagnostic: ts.Diagnostic,
): ReducedDiagnostic {
  if (diagnostic.file) {
    const { line, character } = diagnostic.file.getLineAndCharacterOfPosition(
      diagnostic.start!,
    );
    const message = ts.flattenDiagnosticMessageText(
      diagnostic.messageText,
      '\n',
    );

    return {
      message,
      fileName: path.basename(diagnostic.file.fileName),
      line,
      character,
    };
  } else {
    return {
      message: ts.flattenDiagnosticMessageText(diagnostic.messageText, '\n'),
    };
  }
}

export function getDiagnostics(
  fileNames: string[],
  options: ts.CompilerOptions,
): ReducedDiagnostic[] {
  const program = ts.createProgram(fileNames, options);
  const emitResult = program.emit();

  const diagnostics = ts
    .getPreEmitDiagnostics(program)
    .concat(emitResult.diagnostics);

  return diagnostics.map(reducedDiagnostic);
}

export function diagnosticsForFile(fileName: string) {
  return getDiagnostics([fileName], defaultCompilerOptions);
}
