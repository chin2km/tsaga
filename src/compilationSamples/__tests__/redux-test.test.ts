import { diagnosticsForFile } from '../helpers/helpers';

test('Test forces compiler errors when invalid types are used', async () => {
  const diagnostics = diagnosticsForFile(
    `${__dirname}/../samples/redux-test.ts`,
  );

  expect(diagnostics).toMatchSnapshot();

  // explicit alignment with comments in file, to catch accidental snapshot overwrite
  expect(diagnostics.length).toBe(5);
});

test('Generic types are not correctly inferred (TS 3.4)', async () => {
  const diagnostics = diagnosticsForFile(
    `${__dirname}/../samples/typescript-3.4-inference.ts`,
  );

  expect(diagnostics).toMatchSnapshot();

  // explicit alignment with comments in file, to catch accidental snapshot overwrite
  expect(diagnostics.length).toBe(2);
});
