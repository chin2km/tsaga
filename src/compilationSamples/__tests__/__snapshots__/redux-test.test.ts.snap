// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test forces compiler errors when invalid types are used 1`] = `
Array [
  Object {
    "character": 65,
    "fileName": "redux-test.ts",
    "line": 14,
    "message": "Argument of type '5' is not assignable to parameter of type 'boolean'.",
  },
  Object {
    "character": 12,
    "fileName": "redux-test.ts",
    "line": 15,
    "message": "Argument of type '(url: string | Request, init?: RequestInit | undefined) => Promise<Response>' is not assignable to parameter of type '(_0: undefined, _1: { method: string; body: string; }) => Promise<Response>'.
  Types of parameters 'url' and '_0' are incompatible.
    Type 'undefined' is not assignable to type 'string | Request'.",
  },
  Object {
    "character": 63,
    "fileName": "redux-test.ts",
    "line": 18,
    "message": "Argument of type '404' is not assignable to parameter of type 'Response | Promise<Response>'.",
  },
  Object {
    "character": 17,
    "fileName": "redux-test.ts",
    "line": 20,
    "message": "Argument of type 'true' is not assignable to parameter of type 'string'.",
  },
]
`;
